types/tx.go

```diff
@@ -102,6 +94,232 @@ func ToTxs(txs [][]byte) Txs {
+	// It sorts the new transaction list twice: once when sorting 'all', the total list,
+	// and once by sorting the set of the added, removed, and unmodified transactions indexes,
+	// which, when combined, comprise the complete list of transactions.
+	//
+	// The original list is iterated once and the modified list is iterated multiple times,
+	// one time alongside each of the 3 indexes for a total of 4 iterations of the modified list.
+	// Asymptotically, this yields a total runtime of O(N*log(N) + N + 2*M*log(M) + 4*M),
+	// in the input size of the original list and the input size of the new list respectively.
+	// A 2 * M performance gain is possible but iterating all of the indexes simultaneously
+	// alongside the original list, but the multiple iterations were preferred to be more
+	// readable and maintainable.
+
+	// Sort a copy of the complete transaction slice so we can check for
+	// duplication. The copy is so we do not change the original ordering.
+	// Only the slices are copied, the transaction contents are shared.
+	allCopy := make([]Tx, len(t.all))
+	copy(allCopy, t.all)
+	sort.Sort(Txs(allCopy))
+
+	var size int64
+	for i, cur := range allCopy {
+		size += int64(len(cur))
+		if size > maxSizeBytes {
+			return fmt.Errorf("transaction data size %d exceeds maximum %d", size, maxSizeBytes)
+		}
+
+		// allCopy is sorted, so any duplicated data will be adjacent.
+		if i+1 < len(allCopy) && bytes.Equal(cur, allCopy[i+1]) {
+			return fmt.Errorf("found duplicate transaction with hash: %x", cur.Hash())
+		}
+	}
+
+	// create copies of each of the action-specific indexes so that order of the original
+	// indexes can be preserved.
+	addedCopy := make([]Tx, len(t.added))
+	copy(addedCopy, t.added)
+	sort.Sort(Txs(addedCopy))
+
+	removedCopy := make([]Tx, len(t.removed))
+	copy(removedCopy, t.removed)
+	sort.Sort(Txs(removedCopy))
+
+	unmodifiedCopy := make([]Tx, len(t.unmodified))
+	copy(unmodifiedCopy, t.unmodified)
+	sort.Sort(Txs(unmodifiedCopy))
+
+	// make a defensive copy of otxs so that the order of
+	// the caller's data is not altered.
+	otxsCopy := make([]Tx, len(otxs))
+	copy(otxsCopy, otxs)
+	sort.Sort(Txs(otxsCopy))
+
+	if ix, ok := containsAllTxs(otxsCopy, unmodifiedCopy); !ok {
+		return fmt.Errorf("new transaction incorrectly marked as removed, transaction hash: %x", unmodifiedCopy[ix].Hash())
+	}
+
+	if ix, ok := containsAllTxs(otxsCopy, removedCopy); !ok {
+		return fmt.Errorf("new transaction incorrectly marked as removed, transaction hash: %x", removedCopy[ix].Hash())
+	}
+	if ix, ok := containsNoneTxs(otxsCopy, addedCopy); !ok {
+		return fmt.Errorf("existing transaction incorrectly marked as added, transaction hash: %x", addedCopy[ix].Hash())
+	}
+	return nil
+}
+
+// containsNoneTxs checks that list a contains none of the transactions in list
+// b. If a match is found, the index in b of the matching transaction is returned.
+// Both lists must be sorted.
+func containsNoneTxs(a, b []Tx) (int, bool) {
```

Comment:
I had to read this a couple times to convince myself I had the semantics right. Might it be easier to invert the sense, e.g., `containsAny` vs `containsNone`, and we want false instead of true?

That alone is cosmetic, but I think that change would also let us share a helper between the two cases (`containsdAny` and `containsAll`) to the effect of `contains(tx, txs)`, since then the polarity of the answer would match and the shortcut cases would too. (I think)

Reply:
switched.

Reply:
Per the second part of this:
> That alone is cosmetic, but I think that change would also let us share a helper between the two cases (containsdAny and containsAll) to the effect of contains(tx, txs), since then the polarity of the answer would match and the shortcut cases would too. (I think)

This would blow us up to being `O(n*m)` in the two lists iiuc. Right now, the `containsAnyTxs` and `containsAllTxs` each only iterate through each of their inputs once in the worst case, achieving `O(M+N)` however, a `contains` function would need to, at worst case, iterate the whole list for each input, and wouldn't leverage the fact that the two lists are sorted. Perhaps I'm missing something here though?